ludwig/models/model.py

```diff
@@ -1110,8 +1113,29 @@ def collect_weights(
         pass
 
     def save_weights(self, save_path):
-        # todo tf2: reintroduce functionality
-        #self.weights_save_path = self.saver.save(save_path)
+        # collect all custom metrics used in output features
+        custom_objects = {}
+        for of_name, of in self.ecd.output_features.items():
+            for mfn_name, mfn_obj in of.metric_functions.items():
+                # if module name starts with 'ludwig' this is a custom metric
+                if mfn_obj.__class__.__module__[:6] == 'ludwig':
+                    custom_objects.update(
+                        {mfn_obj.__class__.__name__: mfn_obj.__class__}
+                    )
+        # create pickle of the custom object and save pickled
+        # file in the saved model 'asset' directory
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            custom_objects_fp = os.path.join(tmpdirname, "ludwig_custom_objects.pkl")
+            with open(custom_objects_fp, "wb") as f:
+                pickle.dump(custom_objects, f)
+            # add pickle file to trackable object
+            trackable_obj = tf.train.Checkpoint()
+            custom_objects_asset = tf.saved_model.Asset(custom_objects_fp)
```

Comment:
Here i have a couple questions:
- why do we need to do this? TF Saved model, at least in TF1, was a separate format from the regular save format, so I'm not sure why we are sueing SvedModel functions for this.
- There-s no way to pass the object in memory rather than the pickled file path? This seems clunky, but if there's no other way...

Reply:
I don't know if is just a TF2 maturity issue or not.  Assuming `self.ecd.save()` is the correct api to use in `Model.save_weights()`, one limitation I found is that Ludwig custom metric functions are not recreated on the `tf.keras.models.load_model()`. 

As I understand the documentation, the [`save()` method](https://www.tensorflow.org/api_docs/python/tf/keras/Model#save) creates, by default in TF2, TensorFlow SavedModel format.  This led to [SavedModel documentation](https://www.tensorflow.org/guide/saved_model).  This [section](https://www.tensorflow.org/guide/saved_model#the_savedmodel_format_on_disk) described the file structures used in SavedModel.  From this led to looking up the [concepts of `Assets`](https://www.tensorflow.org/api_docs/python/tf/saved_model/Asset) as a means to save and restore information about Ludwig custom metrics. 

In looking over the documentation, I realized I may be mis-using the `assets` sub-directory to save information about the customer metrics used in the model.  The `assets.extra` directory may be the better place for the custom metrics.  If this is the case, I can rework this section of code to eliminate the use of `tf.train.Checkpoint()` method.

Bottom line:  AFAICT, we need the ability to save information about the ludwig custom metrics to disk on our own and to make it available during model restoration because that is not saved as part of the SavedModel format.

Reply:
Got it. So it seems to me there's a lot of confusion about the formats to save models in TF2 between checkpoints, savedmodel and kers h5... Let's keep on going the direction we have been going so far though.